src/clj/puppetlabs/services/ca/certificate_authority_core.clj

```diff
@@ -56,11 +59,34 @@
       ;; Respond to all CSR validation failures with a 400
       (middleware-utils/plain-response 400 msg))))
 
-(defn handle-get-certificate-revocation-list
-  [{:keys [cacrl]}]
-  (-> (ca/get-certificate-revocation-list cacrl)
-      (rr/response)
-      (rr/content-type "text/plain")))
+  (schema/defn format-http-date :- org.joda.time.DateTime
+    "Formats a http-date into joda time"
+    [http-date :- schema/Str]
+    (time-format/parse
+      (time-format/formatters :rfc822)
+      (string/replace http-date #"GMT" "+0000")))
+
+  (defn handle-get-certificate-revocation-list
+    "Always return the crl if no 'if-modified-since' header is provided or if that header is not in correct http-date format.
+    If the header is present and correctly formatted, only return the crl if the master cacrl is newer than the agent crl."
+    [request {:keys [cacrl]}]
+    (let [agent-crl-last-modified-val (rr/get-header request "If-Modified-Since")
+          agent-crl-last-modified-date-time (if agent-crl-last-modified-val
+                                              (try
+                                                (format-http-date agent-crl-last-modified-val)
+                                                (catch IllegalArgumentException e
+                                                  :invalid-http-date))
+                                              :unknown)
+          master-crl-last-modified-date-time (ca/get-crl-last-modified cacrl)]
+      (if (or (= :unknown agent-crl-last-modified-date-time)
+              (= :invalid-http-date agent-crl-last-modified-date-time)
```

Comment:
so obviously i could just return nil (and check only once for nil) instead of :unknown or :invalid-http-date, but I thought this would be easier to follow/understand.

Reply:
I would be fine with just checking for `nil`, I think, since we don't have any differing behavior between the two cases. Unless you think we might need to add something for one of the cases?